projects/transformers/trainer_mixins/deepspeed.py

```diff
@@ -0,0 +1,129 @@
+
+import torch
+from deepspeed import replace_transformer_layer
+from deepspeed.module_inject import HFBertLayerPolicy
+from transformers import BertLayer
+
+from nupic.torch.modules import SparseWeights
+from projects.transformers.models.sparse_deepspeed_transformer import (
+    SparseDeepSpeedTransformerLayer,
+)
+
+__all__ = ["DeepspeedTransformerLayerMixin", "replace_sparse_transformer_layer"]
+
+
+def replace_sparse_transformer_layer(model, config=None, fp16=True, training=True):
+
+    # Store sparse mask and properties before deepspeed conversion
+    sparsity_per_layer = defaultdict(dict)
+    for i, bert_layer in enumerate(model.encoder.layer):
+        # Only supports huggingface BertLayer
+        assert isinstance(bert_layer, BertLayer)
+
+        # Check if model has sparse bert layers
+        query = bert_layer.attention.self.query
+        if isinstance(query, SparseWeights):
+            # Assume same sparsity for all QKV layers
+            key = bert_layer.attention.self.key
+            value = bert_layer.attention.self.value
+            qkv_mask = torch.cat(
+                (query.zero_mask, key.zero_mask, value.zero_mask), dim=0
+            )
+            sparsity_per_layer[i]["attn_qkv"] = (query.sparsity, qkv_mask)
+
+        attn_out = bert_layer.attention.output.dense
+        if isinstance(attn_out, SparseWeights):
+            sparsity_per_layer[i]["attn_out"] = (
+                attn_out.sparsity,
+                attn_out.zero_mask,
+            )
+
+        intermediate = bert_layer.intermediate.dense
+        if isinstance(intermediate, SparseWeights):
+            sparsity_per_layer[i]["inter"] = (
+                intermediate.sparsity,
+                intermediate.zero_mask,
+            )
+
+        output = bert_layer.output.dense
+        if isinstance(output, SparseWeights):
+            sparsity_per_layer[i]["output"] = (output.sparsity, output.zero_mask)
+
+    # Convert transform layers to deepspeed.
+    ds_model = replace_transformer_layer(
+        BertLayer,
+        model,
+        policy=HFBertLayerPolicy,
+        fp16=fp16,
+        config=config,
+        training=training,
+        encoder_decoder=True,
+    )
+
+    # This conversion replaces all HF BertLayer with DeepSpeedTransformerLayer.
+    # In consequence any SparseWeight layer manually injected into the original
+    # BertLayer will be replaced with a regular nn.Linear layer. For this reason
+    # we need to convert all DeepSpeedTransformerLayer to
+    # SparseDeepSpeedTransformerLayer and restore the original zero mask and
+    # sparsity configuration
+    for i, sparse_config in sparsity_per_layer.items():
+        ds_layer = ds_model.encoder.layer[i]
+
+        attn_qkv_sparsity, attn_qkv_mask = sparse_config["attn_qkv"]
+        attn_out_sparsity, attn_out_mask = sparse_config["attn_out"]
+        inter_sparsity, inter_mask = sparse_config["inter"]
+        output_sparsity, output_mask = sparse_config["output"]
+
```

Comment:
You don't use the `attn_out_sparsity`, `inter_sparsity`, and `output_sparsity` it seems. Should you just not define them so it's clear that the sparsity of `attn_qkv_sparsity` will be used for the module?